ACP-83: change fee update algorithm (#97)
avalanche-foundation/ACPs: ACPs/83-dynamic-multidimensional-fees/README.md

## Specification

We introduce the multidimensional scheme first and then how to apply the dynamic fee update rule for each fee dimension. Finally we list the new block verification rules, valid once the new fee scheme activates.

### Multidimensional scheme components

We define four fee dimensions, `Bandwidth`, `DBReads`, `DBWrites`, `Compute`, to describe transactions complexity. In more details:

- `Bandwidth` measures the transaction size in bytes, as encoded by the Avalanchego codec. Byte length is a proxy for the network resources needed to disseminate the transaction.
- `DBReads` measures the number of DB reads needed to verify the transactions. DB reads include UTXOs reads and any other state quantity relevant for the specific transaction.
- `DBWrites` measures the number of DB writes following the transaction verification. DB writes include UTXOs generated as outputs of the transactions and any other state quantity relevant for the specific transaction.

> - `Compute` measures the number of signatures to be verified, including UTXOs ones and those related to authorization of specific operations.

Reviewer: Could we simplify these? The explanations seem longer than necessary.

Author: This idea is to try and be precise, so to let users replicate the measurements I did in the complexities repo, should they want to.
I think the details were asked by @dhrubabasu and I'll pull him in to ask his opinion

### Dynamic scheme components

Fee rates are updated in time, to allow a fee increase when network is getting congested. Each new block is a potential source of congestion, as its transactions carry complexity that each validator must process to verify and eventually accept the block. The more complexity carries a block, and the more rapidly blocks are produced, the higher the congestion.  

> We seek a scheme that rapidly increases the fees when blocks complexity goes above a defined threshold and that equally rapidly decreases the fees once complexity goes down (because blocks carry less/simpler transactions, or because they are produced more slowly). We define the desired threshold as a *target complexity rate* $T_i$: we would want to process every second a block whose cumulated complexity is $T_i$. Any complexity more than that causes some congestion that we want to penalize via fees; any congestion below $T_i$ causes the network to be under utilized.

Reviewer: This is not exactly true that we want the fees to equally increase/decrease when block complexity is above/below target. Assuming that's multiplicative, then we'd end up with the same problem of long term usage being above target as in EIP-1559 (https://ethresear.ch/t/make-eip-1559-more-like-an-amm-curve/9082).

Could we re-phrase the goal here?

> In order to update fees rates we will track, per each block and each fee dimension, a parameter called cumulated excess complexity. Fee rates applied to a block will be defined in terms of cumulated excess complexity as we show in the following.

Reviewer: Should this be cumulative instead of cumulated?

Let's define cumulated excess complexity first.

Suppose that a block $B_t$ is the current chain tip. $B_t$ has the following features:

- $t$ is its timestamp.
- $Excess_{i,t}$ is the cumulated excess complexity along fee dimension $i$.

Say a new block $B_{t + \Delta T}$ is accepted on top of $B$, with the following features:

- $t + \Delta T$ is its timestamp
- $B_i$ is its complexity.

Then cumulated excess complexity once $B_{t + \Delta T}$ is accepted is updated as follow:

> $$Excess_{i,t + \Delta T} = max\large(0, Excess_{i,t} - T_i \times \Delta T\large) + B_i$$

Reviewer: Could we re-organize this so that the formula goes first?

Since each term is fairly self-explanatory, I think this will make it more readable.

Author: Indeed, done

Finally let's define the fee rates update formula.

Say block $B_t$ is chain tip and a block $B_{t + \Delta T}$ is incoming on top of it. Then

> $$ r_{i,t + \Delta T} = r^{min}_i \times e^{\frac{Excess_{i,t} - T_i \times \Delta T}{Denom_i}} $$

Reviewer 2: this equation isn't rendered correctly in the preview, not 100% sure the suggestion will change that  
```suggestion
$$r_{i,t + \Delta T} = r^{min}_i \times e^{\frac{Excess_{i,t} - T_i \times \Delta T}{Denom_i}}$$
```

Author: Thanks for spotting this. I have simplified the formula (dropping some subscripts) to fix the rendering and it should be good now.

where

- $r_{i,t + \Delta T}$ is the fee rate applied to block $B_{t + \Delta T}$ along fee dimension $i$
- $r^{min}_i$ is the minimal fee rate along fee dimension $i$
- $Denom_i$ is a normalization constant for the fee dimension $i$

> The update formula guarantees that fee rates increase if incoming blocks are complex (large $B_i$) and if blocks are emitted rapidly (small $\Delta T$). Simmetrically fee rates decrease to the minimum if incoming blocks are less complex and if blocks are produced less frequently.  

Reviewer: ```suggestion
The update formula guarantees that fee rates increase if incoming blocks are complex (large $B_i$) and if blocks are emitted rapidly (small $\Delta T$). Symmetrically, fee rates decrease to the minimum if incoming blocks are less complex and if blocks are produced less frequently.  
```

- **Bound block complexity**. For each fee dimension $i$, we define a *maximal block complexity* $C_i$. A block is only valid if the block complexity $B_i$ is less than the maximum block complexity: $B_i \leq C_i$.
- **Verify transaction fee**. When verifying each transaction in a block, we confirm that it can cover its own base fee. Note that both base fee and optional priority fees are burned.

## User Experience

### How will the wallets estimate the fees?

> AvalancheGo nodes will provide new APIs exposing the current and expected fee rates, as they are likely to change block by block. Wallets can then use the fees rates to select UTXOs to pay the transaction fees. Moreover Avalanchego implementation proposed above offers a `fees.Calculator` struct that can be reused by wallets and downstreams to evaluate calculate fees.

Reviewer: ```suggestion
AvalancheGo nodes will provide new APIs exposing the current and expected fee rates, as they are likely to change block by block. Wallets can then use the fees rates to select UTXOs to pay the transaction fees. Moreover, the Avalanchego implementation proposed above offers a `fees.Calculator` struct that can be reused by wallets and downstream projects to evaluate calculate fees.
```

### How will wallets be able to re-issue Txs at a higher fee?

> Wallets should be able to simply re-issue the transaction since current AvalancheGo implementation drops mempool transactions whose fee rate is lower than current one. More specifically, a transaction may be valid the moment it enters the mempool and it won’t be re-verified as long as it stays in there. However, as soon as the transaction is selected to be included in the next block, it is re-verified against the latest preferred tip. If fees are not enough by this time, the transaction is dropped and the wallet can simply re-issue it at a higher fee, or wait for the fee rate to go down. Note that priority fees offer some buffer space against an increase in the fee rate. A transaction paying just the base fee will be evicted from the mempool in the face of a fee rate increase, while a transaction paying some extra priority fee may have enough buffer room to stay valid after some amount of fee increase.

Reviewer: Can we be consistent with usage of either Avalanchego or AvalancheGo? I prefer AvalancheGo personally
